pyomo/opt/results/results_.py

```diff
@@ -75,18 +78,38 @@ def _repn_(self, option):
         return tmp
 
     def write(self, **kwds):
-        if 'filename' in kwds:
-            OUTPUT = open(kwds['filename'],"w")
-            del kwds['filename']
-            kwds['ostream']=OUTPUT
-            self.write(**kwds)
-            OUTPUT.close()
-            return
-
-        if not 'format' in kwds or kwds['format'] == 'yaml':
-            self.write_yaml(**kwds)
+        _fmt = kwds.pop('format', None)
+        if _fmt:
+            _fmt = _fmt.lower()
+        fname = kwds.pop('filename', None)
+
+        if fname:
+            ext = os.path.splitext(fname)[1].lstrip('.')
+            normalized_ext = {
+                'json': 'json',
+                'jsn': 'json',
+                'yaml': 'yaml',
+                'yml': 'yaml',
+            }.get(ext, None)
+            if not _fmt:
+                _fmt = normalized_ext
+            elif normalized_ext and _fmt != normalized_ext:
+                logger.warning(
+                    "writing results to file (%s) using what appears "
+                    "to be an incompatible format (%s)" % (fname, _fmt))
+            with open(fname, "w") as OUTPUT:
+                kwds['ostream'] = OUTPUT
+                kwds['format'] = _fmt
+                self.write(**kwds)
         else:
-            self.write_json(**kwds)
+            if not _fmt:
+                _fmt = 'yaml' if yaml_available else 'json'
```

Comment:
@jsiirola we can write a yaml file without yaml being available and some of our tests rely on this behavior. I'm conflicted over removing the `if yaml_available else 'json'` part of this line to match the previous behavior or skipping 3 tests when yaml is not available.

Any preference?